Task 03 - Implement HTML markup using components

nikolay-borzov opened the pull request:
resolves #4 

- To test Error Boundaries add `?throwError=1` to the URL

Commit:
feat: Layout components (WIP)

Commit:
feat: Layout (WIP)

- header, footer
- search form
- empty search results

Commit:
feat (task_03): Search results (WIP)

- Search results panel
- Movies found count

nikolay-borzov committed:
feat: Layout components (WIP)

- Radio `component`
- Films grid components

Commit:
chore(merge): master

Commit:
feat(task_03): Films grid

- Add `ErrorBoundary` (initial)
- Simple API integration

nikolay-borzov committed:
feat(task_03): Error Boundary

- Add polyfill for `URLSearchParams`
- Add loading indicator
- WIre up `react-hot-loader`, `babel-polyfill`

Commit:
feat(task_03): Sorting

- Add `react-load-image` and `react-toastify`
- Simplify webpack dev config: move hmr to the end
- FInish `Radio` component
- Implement sorting (doesn't work on server-side)

tsvetkovv commented on src/search/search-container.jsx:
```diff
@@ -40,15 +53,45 @@ export class SearchContainer extends React.Component {
+    })
+  }
+
+  onSortByChange(sortBy) {
```
You can define class field to avoid using `bind(this)` in the render method
`onSortByChange = (sortBy) => {`

tsvetkovv replied:
Are you going to split the css file?

nikolay-borzov replied:
> You can define class field to avoid using bind(this) in the render method
onSortByChange = (sortBy) => {

Yeah, I'll probably follow this style

> Are you going to split the css file?

Eventually yes. I'm planning to extract component specific styles to their own `styles.css`. I consider changing project structure

This task is still WIP

nikolay-borzov committed:
fix(task_03): Fix sorting

nikolay-borzov committed:
feat(task_03): Change project structure

- Move component specific styles to a separate files
- Add `jsconfig.json` for VSCode
- Add `limit` quesry string param
- Implement search by title/genre
- Add `filmService`

tsvetkovv commented on src/components/films-grid-item.jsx:
```diff
@@ -0,0 +1,44 @@
+  constructor(props) {
+    super(props)
+
+    this.film = this.props.film
```
Why do you bind to class field? If `this.props` will change, the component doesn't update (constructor is not called when props are changed)
Let's look at example:
I modified Films Grid component.
```JSX
export class FilmsGrid extends React.PureComponent {
  static propTypes = {
    films: PropTypes.arrayOf(PropTypes.object).isRequired
  }

  films = [{
      "id": 447365,
      "title": "Guardians of the Galaxy Vol. 3",
      "tagline": "",
      "vote_average": 0,
      "vote_count": 9,
      "release_date": "2020-05-01",
      "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
      "overview": "The third film based on Marvel's Guardians of the Galaxy.",
      "budget": 0,
      "revenue": 0,
      "genres": ["Action", "Adventure", "Science Fiction"],
      "runtime": null
  }, {
      "id": 424785,
      "title": "Transformers 7",
      "tagline": "",
      "vote_average": 0,
      "vote_count": 0,
      "release_date": "2019-06-26",
      "poster_path": "https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg",
      "overview": "Plot unknown.",
      "budget": 0,
      "revenue": 0,
      "genres": ["Science Fiction", "Action", "Adventure"],
      "runtime": null
  }, {
      "id": 299534,
      "title": "Untitled Avengers",
      "tagline": "",
      "vote_average": 0,
      "vote_count": 9,
      "release_date": "2019-04-24",
      "poster_path": "https://image.tmdb.org/t/p/w500/aMETsaNNcDc6g5ZatQtVbySnSaA.jpg",
      "overview": "The culmination of the Marvel Cinematic Universe.",
      "budget": 0,
      "revenue": 0,
      "genres": ["Action", "Adventure", "Science Fiction"],
      "runtime": null
  }]

  state = {
    currentFilmIndex: 0
  }

  componentDidMount() {
    setInterval(() => {
      this.setState({
        currentFilmIndex: (this.state.currentFilmIndex + 1) % this.films.length
      })
    }, 1000)
  }

  render() {
    const film = this.films[this.state.currentFilmIndex];

    return (
      <div className="film-grid">
        <FilmsGridItem film={film} />
      </div>
    )
  }
}
```

It emits new film every 1 sec. `film` prop of `FilmGridItem` changes, but updates don't render.
There is demo how it works:
![2018-04-27_21-34-27](https://user-images.githubusercontent.com/11594423/39376850-26784452-4a64-11e8-93de-3b1ff6842e1f.gif)

nikolay-borzov replied:
I see. But it's unlikely that a film data will change. I most cases a new FilmsGridItem is created.
I'll change a class field to props one.

tsvetkovv replied:
I just think you missed that props can be changed, instead or replacing of whole component instance. You can face with implicit bug in future

nikolay-borzov replied:
Agree

tsvetkovv commented on src/pages/search/components/search-form.jsx:
```diff
@@ -0,0 +1,93 @@
+
+import { Radio } from '../../../components'
+
+export class SeachForm extends React.PureComponent {
```
mistype in `SeachForm`

nikolay-borzov replied:
Good catch! I have a problem with this word all the time.

tsvetkovv commented on src/pages/search/components/search-form.jsx:
```diff
@@ -0,0 +1,93 @@
+
+    this.state = {
+      search: this.props.search,
+      searchBy: this.props.searchBy
```
It's not good idea to duplicate props in state, cause you are supposed to synchronize the state when props are changed

nikolay-borzov replied:
I know I read [about it](https://reactjs.org/docs/react-component.html#constructor).
I didn't know the other way to pass changed data on form submit. I don't want to pass `search` and `searchBy` values via separate callbacks. Also, it's supposed to be onetime initialization.

On second thought, I think I can override `getDerivedStateFromProps`, so that component state will be in sync with props. Will it work here?

tsvetkovv replied:
Good article and definition state as "forks of props" for this case, I haven't been read this docs a long time.
Why do you think parent component won't want to change props? 
To be honest I don't understand why do you you local state in the component. I think it's enough to use [uncontrolled component](https://reactjs.org/docs/uncontrolled-components.html) without state. What do you think?

nikolay-borzov replied:
I see search params data flow as:
1. The page component is loaded - it get's initial search params from URL query string
2. The component passes search parameters to it's children components
3. Child components set their inputs values from props
4. If a component contains a form element it accumulates inputs' data and sends data to parent on submit
5. In general, child components sends data to parent when data changed via user input
6. The page component updates URL query string and update

I think the parent shouldn't change search params to avoid erasing user's input

Using uncontrolled components looks cumbersome to me. But I will ty to implement logic this way

nikolay-borzov committed:
fix(task_03): Fixes after code review

- Turn presentational components into functions
- Implement uncontrolled component  behavior for `Radio` component

nikolay-borzov committed:
chore(deps): Update

nikolay-borzov committed:
feat(task_03): Movie page (initial)

- Implement `loading-block`, `toast-error` components
- Add `cssnext` and `postcss-import`

nikolay-borzov committed:
feat(task_03) Films page

- Implement `content-image` component

tsvetkovv commented on src/app.jsx:
```diff
@@ -0,0 +1,31 @@
+  const searchParams = new URLSearchParams(location.search)
+
+  // TODO: Remove after React Router integration
+  const isFilmPage = searchParams.has('id')
```
Good workaraund 😀

tsvetkovv reviewed (approved):
Well done
